### PR TITLE
Simplify Struct Set/Delete

### PR DESCRIPTION
**Pull request by arv**
The old code was hard to follow and it ended up encoding twice when a
new field was added.

### CODE REVIEW

File: go/types/struct.go [hunk 1]
```diff
@@ -270,57 +343,37 @@ func (s Struct) Get(n string) Value {
...
-func (s Struct) set(w *binaryNomsWriter, n string, v Value, addedCount int) Struct {
-	// TODO: Reuse bytes if we end up adding a field
-	dec := s.decoder()
-	StructKind.writeTo(w)
-	dec.skipKind()
-	dec.copyString(w)
-	count := dec.readCount()
-	w.writeCount(count + uint64(addedCount))
+func (s Struct) set(name string, value Value) Struct {
+	entries := s.structEntries()
```
Comment (arv) [inline go/types/struct.go]: This could be done without allocating this slice but this way is much cleaner.

**Comment by arv** [comment]
Ping

File: go/types/struct.go [hunk 2]
```diff
@@ -271,56 +271,58 @@ func (s Struct) Get(n string) Value {
...
+	w.writeString(n)
+	v.writeTo(&w)
+	w.writeRaw(tail)
+
+	return Struct{s.vrw, w.data()}
 }
 
-func (s Struct) set(w *binaryNomsWriter, n string, v Value, addedCount int) Struct {
-	// TODO: Reuse bytes if we end up adding a field
+// splitFieldsAt splits the buffer into two parts. The fields comming before the field we are looking for
```
Comment (rafael-atticlabs) [inline go/types/struct.go]: `coming`

File: go/types/struct.go [hunk 3]
```diff
@@ -331,32 +333,18 @@ func (s Struct) IsZeroValue() bool {
...
-		} else {
-			w.writeString(name)
-			dec.copyValue(&w)
-		}
-	}
+	prolog, head, tail, count, found := s.splitFieldsAt(n)
 
+	w := binaryNomsWriter{make([]byte, len(s.buff)), 0}
+	w.writeRaw(prolog)
 	if found {
```
Comment (rafael-atticlabs) [inline go/types/struct.go]: shouldn't this just `return s if !found`?

Comment (arv) [reply go/types/struct.go]: yup
